File: APIO/20/practice/party/java/grader.java
public class grader {
  private static int N, M, Q;

  private static long[] maleAcquaintances;
  private static long[] femaleAcquaintances;

  private static void wrongAnswer(String message) {
    System.out.printf("WA: %s\n", message);
    System.exit(0);
  }

  public static int[] ask(int[] G) {
    if (--Q < 0) {
      wrongAnswer("Too many queries");
    }
    long males = 0;
    long females = 0;

    int previousIndex = -1;
    for (int i = 0; i < G.length; ++i) {
      if (G[i] <= previousIndex || G[i] >= N + M) {
        wrongAnswer("Invalid index");
      }
      if (G[i] < N) {
        females |= maleAcquaintances[G[i]];
      } else {
        males |= femaleAcquaintances[G[i] - N];
      }
    }

    int[] result = new int[Long.bitCount(males) + Long.bitCount(females)];
    int currentIndex = 0;
    while (males > 0) {
      int maleIndex = Long.numberOfTrailingZeros(males);
      result[currentIndex++] = maleIndex;
      males ^= 1L << maleIndex;
    }
    while (females > 0) {
      int femaleIndex = Long.numberOfTrailingZeros(females);
      result[currentIndex++] = femaleIndex + N;
      females ^= 1L << femaleIndex;
    }
    return result;
  }

  public static void main(String[] args) {
    InputReader inputReader = new InputReader(System.in);
    N = inputReader.readInt();
    M = inputReader.readInt();
    Q = inputReader.readInt();
    maleAcquaintances = new long[N];
    femaleAcquaintances = new long[M];

    for (int i = 0; i < N + M; ++i) {
      String buffer = inputReader.readString();
      for (int j = 0; j < N + M; ++j) {
        if (buffer.charAt(j) == '1') {
          if (i < N) {
            maleAcquaintances[i] |= 1L << (j - N);
          } else {
            femaleAcquaintances[i - N] |= 1L << j;
          }
        }
      }
    }

    party solver = new party();
    int unusualGuest = solver.findUnusualGuest(N, M, Q);

    if (unusualGuest < 0 || unusualGuest >= N + M) {
      wrongAnswer("Invalid index");
    }
    if (unusualGuest < N) {
      if (Long.bitCount(maleAcquaintances[unusualGuest]) == 1) {
        wrongAnswer("Non-unusual guest found");
      }
    } else {
      if (Long.bitCount(femaleAcquaintances[unusualGuest - N]) == 1) {
        wrongAnswer("Non-unusual guest found");
      }
    }

    System.out.println(unusualGuest);
  }
}


class InputReader implements java.io.Closeable {
  private java.io.InputStream stream;
  private byte[] buf = new byte[1024];
  private int curChar;
  private int numChars;

  public InputReader(java.io.InputStream stream) {
    this.stream = stream;
  }
  
  @Override
  public void close() {
    try {
      this.stream.close();
    } catch (java.io.IOException e) {
    }
  }

  public int read() {
    if (numChars == -1) {
      throw new java.util.InputMismatchException();
    }
    if (curChar >= numChars) {
      curChar = 0;
      try {
        numChars = stream.read(buf);
      } catch (java.io.IOException e) {
        throw new java.util.InputMismatchException();
      }
      if (numChars <= 0) {
        return -1;
      }
    }
    return buf[curChar++];
  }

  public int readInt() {
    int c = eatWhite();
    int sgn = 1;
    if (c == '-') {
      sgn = -1;
      c = read();
    }
    int res = 0;
    do {
      if (c < '0' || c > '9') {
        throw new java.util.InputMismatchException();
      }
      res *= 10;
      res += c - '0';
      c = read();
    } while (!isSpaceChar(c));
    return res * sgn;
  }

  public String readString() {
    int c = eatWhite();
    StringBuilder res = new StringBuilder();
    do {
      if (Character.isValidCodePoint(c))
        res.appendCodePoint(c);
      c = read();
    } while (!isSpaceChar(c));
    return res.toString();
  }

  public String readLine() {
    StringBuilder res = new StringBuilder();
    while (true) {
      int c = read();
      if (c == '\n' || c == '\r' || c == -1)
        break;
      if (Character.isValidCodePoint(c))
        res.appendCodePoint(c);
    }
    return res.toString();
  }

  private int eatWhite() {
    int c = read();
    while (isSpaceChar(c)) {
      c = read();
    }
    return c;
  }

  public static boolean isSpaceChar(int c) {
    return c == ' ' || c == '\n' || c == '\r' || c == '\t' || c == -1;
  }
}

File: contests/APIO-2020/practice/party/cpp/grader.cpp
#include "party.h"

#include <cassert>
#include <cstdio>

#include <string>
#include <vector>

static int N, M, Q;

static std::vector<long long> male_acquaintances;
static std::vector<long long> female_acquaintances;

static void wrongAnswer(std::string message) {
  printf("WA: %s\n", message.c_str());
  exit(0);
}

std::vector<int> ask(std::vector<int> G) {
  if (--Q < 0) {
    wrongAnswer("Too many queries");
  }
  long long males = 0;
  long long females = 0;

  int previous_index = -1;
  for (int i = 0; i < G.size(); ++i) {
    if (G[i] <= previous_index || G[i] >= N + M) {
      wrongAnswer("Invalid index");
    }
    if (G[i] < N) {
      females |= male_acquaintances[G[i]];
    } else {
      males |= female_acquaintances[G[i] - N];
    }
  }

  std::vector<int> result;
  while (males > 0) {
    int male_index = __builtin_ctzll(males);
    result.push_back(male_index);
    males ^= 1LL << male_index;
  }
  while (females > 0) {
    int female_index = __builtin_ctzll(females);
    result.push_back(female_index + N);
    females ^= 1LL << female_index;
  }
  return result;
}

int main() {
  assert(3 == scanf("%d %d %d", &N, &M, &Q));
  male_acquaintances.resize(N);
  female_acquaintances.resize(M);

  for (int i = 0; i < N + M; ++i) {
    char buffer[N + M + 1];
    assert(1 == scanf("%s", buffer));
    for (int j = 0; j < N + M; ++j) {
      if (buffer[j] == '1') {
        if (i < N) {
          male_acquaintances[i] |= 1LL << (j - N);
        } else {
          female_acquaintances[i - N] |= 1LL << j;
        }
      }
    }
  }

  int unusual_guest = findUnusualGuest(N, M, Q);

  if (unusual_guest < 0 || unusual_guest >= N + M) {
    wrongAnswer("Invalid index");
  }
  if (unusual_guest < N) {
    if (__builtin_popcountll(male_acquaintances[unusual_guest]) == 1) {
      wrongAnswer("Non-unusual guest found");
    }
  } else {
    if (__builtin_popcountll(female_acquaintances[unusual_guest - N]) == 1) {
      wrongAnswer("Non-unusual guest found");
    }
  }

  printf("%d\n", unusual_guest);
  
  return 0;
}
